src/main/scala/org/broadinstitute/dsde/workbench/sam/google/GoogleExtensions.scala

```diff
@@ -228,44 +228,41 @@ class GoogleExtensions(val directoryDAO: DirectoryDAO, val accessPolicyDAO: Acce
     // The normal situation is that the pet either exists in both ldap and google or neither.
     // Sometimes, especially in tests, the pet may be removed from ldap, but not google or the other way around.
     // This code is a little extra complicated to detect the cases when a pet does not exist in google, ldap or both
     // and do the right thing.
-    val pet = for {
-      maybeServiceAccount <- googleIamDAO.findServiceAccount(project, petSaName)
+    val createPet = for {
+      maybeServiceAccount <- IO.fromFuture(IO(googleIamDAO.findServiceAccount(project, petSaName)))
       maybePet <- directoryDAO.loadPetServiceAccount(PetServiceAccountId(user.id, project))
 
       serviceAccount <- maybeServiceAccount match {
         // SA does not exist in google, create it and add it to the proxy group
-        case None => for {
-          sa <- googleIamDAO.createServiceAccount(project, petSaName, petSaDisplayName)
-          r <- withProxyEmail(user.id){proxyEmail => googleDirectoryDAO.addMemberToGroup(proxyEmail, sa.email)}
-        } yield sa
-
+        case None =>
+          for {
+            sa <- IO.fromFuture(IO(googleIamDAO.createServiceAccount(project, petSaName, petSaDisplayName)))
+            r <- IO.fromFuture(IO(withProxyEmail(user.id){proxyEmail => googleDirectoryDAO.addMemberToGroup(proxyEmail, sa.email)}))
+          } yield sa
         // SA already exists in google, use it
-        case Some(sa) => Future.successful(sa)
+        case Some(sa) => IO.pure(sa)
       }
       pet <- (maybePet, maybeServiceAccount) match {
         // pet does not exist in ldap, create it and enable the identity
-        case (None, _) => for {
-          p <- directoryDAO.createPetServiceAccount(PetServiceAccount(PetServiceAccountId(user.id, project), serviceAccount))
-          _ <- directoryDAO.enableIdentity(p.id)
-        } yield p
-
+        case (None, _) =>
+          for {
+            p <- directoryDAO.createPetServiceAccount(PetServiceAccount(PetServiceAccountId(user.id, project), serviceAccount))
+            _ <- directoryDAO.enableIdentity(p.id)
+          } yield p
         // pet already exists in ldap, but a new SA was created so update ldap with new SA info
-        case (Some(p), None) => directoryDAO.updatePetServiceAccount(p.copy(serviceAccount = serviceAccount))
+        case (Some(p), None) =>
+          directoryDAO.updatePetServiceAccount(p.copy(serviceAccount = serviceAccount))
 
         // everything already existed
-        case (Some(p), Some(_)) => Future.successful(p)
+        case (Some(p), Some(_)) => IO.pure(p)
       }
     } yield pet
 
-    // In the case of high concurrency, the above code may find that there is no pet SA and then try to create one, getting
-    // a 409 Conflict in Google if another call has since created it. This recoverWith will catch that case and return the
-    // newly-created pet SA instead of simply failing.
-    pet recoverWith {
+    createPet.unsafeToFuture() recoverWith {
```

Comment:
keep the comment!

Reply:
It'll be removed in the lock PR. so I'll leave this as is